tools/single_use/fix_double_escape.php

```diff
@@ -116,10 +119,70 @@
 
 //SECOND loop, depending on flags, report or fix values.
 if ($errorsDetected) {
-    //TODO CODE FOR CHECKING FIELD MAX LENGTH AND COMPARING TO ACTUAL LENGTH IN ORDER TO BUILD THE CONFIRMEDTRUNCATIONS array defined above
-
-    if ($useHistory) {
-        // TODO CODE FOR CHECKING HISTORY TABLE AND BUILDING BETTER REPAIR SUGGESTIONS
+	foreach ($escapedFields as $tableName => $fieldsArray) {
+		$fieldsList = "'" . implode($fieldsArray, "','") . "'";
+
+		$maxFieldLengths[$tableName] = $DB->pselectWithIndexKey(
+			"SELECT COLUMN_NAME, CHARACTER_MAXIMUM_LENGTH
+			FROM INFORMATION_SCHEMA.COLUMNS 
+			WHERE TABLE_NAME=:tbl AND COLUMN_NAME IN ($fieldsList)",
+			array('tbl' => $tableName),
+			'COLUMN_NAME'
+		);
+	}
+
+	// Start comparing the value length of the escaped entry to the character maximum for that field
+	// In order to begin identifying cases of truncation
+	foreach ($escapedEntries as $tableName => $rows) {
+		foreach ($rows as $cid => $entries) {
+			foreach ($entries as $field => $value) {
+				if (strlen($value) == $maxFieldLengths[$tableName][$field]['CHARACTER_MAXIMUM_LENGTH']) {
+					$confirmedTruncations[$tableName][$cid][$field] = $value;
+					$CIDs[] = $cid;
+					$allFields[] = $field;
+					$allTables[] = $tableName;
+				}
+			}
+		}
+	}
+
+if ($useHistory) {
+	// Construct history query
+	$listCIDs = "'" . implode($CIDs, "','") . "'";
+	$listFields = "'" . implode($allFields, "','") . "'";
+	$listTables = "'" . implode($allTables, "','") . "'";
+	$query = "SELECT primaryVals as CommentID, col, tbl, new, changeDate
+		FROM history 
+		WHERE primaryCols='CommentID' AND
+		primaryVals IN ($listCIDs) AND
+		tbl IN ($listTables) AND
+		col IN ($listFields)";	
+	
+	$result = $DB->pselect(
+		$query, 
+		[]
+	);
+
+	// Reverse the array obtained from history data so that it is in anti-chronological order
+	// Does not preserve keys, assuming that the keys are indexes & not relevant info
+	$historyData = array_reverse($result, false);
```

Comment:
@jesscall there is no garantee that the array returned in in chronological order. you should instead use an order by in the query itself